File: src/Java/algorithms/arrandstring/RemoveDuplicatesFromSortedArrayII.java
package algorithms.arrandstring; 

import java.util.*;

public class RemoveDuplicatesFromSortedArrayII {
    /* Function: Each dup can appear at most twice.  
     * Solution: Use set to mark, use len as pointer. 
    */
    public int removeDuplicates(int[] A) {
        if (A == null || A.length == 0) {
            return 0; 
        }
        
        int len = 0; 
        HashSet<Integer> set = new HashSet<Integer>(); 
        for (int i = 0; i < A.length; ++i) {
            if (!set.contains(A[i])) {
                if (i != 0 && A[i] == A[i - 1]) {
                    set.add(A[i]); 
                }
                A[len++] = A[i]; 
            }
        }
        
        return len; 
    }
}

File: src/CPP/src/leetcode/RemoveDuplicatesFromSortedArrayII.cpp
/*
 * [Source] https://leetcode.com/problems/remove-duplicates-from-sorted-array-ii/
 * [Difficulty]: Medium
 * [Tag]: Array
 * [Tag]: Two Pointers
 */

#include <iostream>

using namespace std;

// [Solution]: Use a cnt. When find current char is same as the slow pointer, increase cnt. If cnt reach the number, reset it and move forward
// [Corner Case]:
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/arrandstring/RemoveDuplicatesFromSortedArrayII.java
 */

/* Java solution
public class Solution {
    public int removeDuplicates(int[] A) {
        HashMap<Integer, Integer> map=new HashMap<Integer, Integer>();
		ArrayList<Integer> list=new ArrayList<Integer>();
		for(int i=0; i<A.length; i++){
			if(!map.containsKey(A[i])){
				map.put(A[i], 1);
				list.add(A[i]);
			}else if(map.get(A[i])<2){
				map.put(A[i], map.get(A[i])+1);
				list.add(A[i]);
			}
		}
		for(int i=0; i<list.size(); i++){
			A[i]=list.get(i);
		}
		return list.size();
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
